### PR TITLE
Furlough

### LINKED ISSUE #242: Integrate furlough
- [x] Create db table for furlough
- [x] Create simple CRUD 
  - [x] Print table
  - [x] Only print furloughs table if there is any furloughs for this work relation
  - [x] Separate CRUD for text, separate place similar to Avdelinger
  - [x] Select instead of input for text
- [x] Add it to XML builder for altinn report
- [x] Only add furlough to the report if the furlough is in that report's period
- [x] Validate
  - [x] Text needs to be a string, can't be empty
    - [x] Report bulder
    - [x] Employee edit page
  - [x] Start date needs to be a valid date
    - [x] Report bulder
    - [x] Employee edit page
  - [x] End date needs to be a valid date and not before the start date
    - [x] Report bulder
    - [x] Employee edit page
  - [x] Percentage, must be higher than 0 up to 100, range (0, 100]
    - [x] Report bulder
    - [x] Employee edit page
  - [x] Type needs to be a string, can't be empty
    - [x] Report bulder
    - [x] Employee edit page
  - [x] Print validation errors on employee edit page
  - [x] Refuse to send altinn report if any of the fields are missing

### PR DESCRIPTION
**Pull request by mladjo2505**
Closes #242

### CODE REVIEW

File: modules/accountplan/view/record.inc [hunk 1]
```diff
@@ -301,15 +301,28 @@ elseif($_lib['input']->getProperty('action_work_relation_delete')) {
   $workrelation_ids = $_POST['work_relations_to_delete'];
   if (!empty($workrelation_ids)) {
     $query = sprintf("DELETE FROM workrelation WHERE WorkRelationID IN (%s)", implode($workrelation_ids, ', '));
+    # TODO delete workrelationfurlough
```
Comment (stoivo) [inline modules/accountplan/view/record.inc]: TODO

File: modules/accountplan/view/employee.php [hunk 2]
```diff
@@ -470,6 +574,7 @@
 }
 ?>
 </table>
+<a href="<? print $_lib['sess']->dispatch ?>t=furlough.textlist">Legg til ny permisjons og permitterings tekst</a>
```
Comment (mladjo2505) [inline modules/accountplan/view/employee.php]: Should be `Legg til ny permisjon og permittering tekst`

File: modules/furlough/view/textlist.php [hunk 3]
```diff
@@ -0,0 +1,57 @@
+<?
+require_once "record.inc";
+
+$query_furloughtexts  = "select * from furloughtext order by FurloughTextID";
+$result_furloughtexts = $_lib['db']->db_query($query_furloughtexts);
+
+print $_lib['sess']->doctype ?>
+
+<head>
+    <title>Empatix - Permisjon og permittering text liste</title>
```
Comment (mladjo2505) [inline modules/furlough/view/textlist.php]: should be `tekst`

File: modules/furlough/view/textlist.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+<body>
+
+<? includeinc('top') ?>
+<? includeinc('left') ?>
+<? print $_lib['message']->get() ?>
+
+<table class="lodo_data">
+<thead>
+  <tr>
+    <th colspan="2">Permisjon og permittering texter:</th>
```
Comment (mladjo2505) [inline modules/furlough/view/textlist.php]: should be `tekster`

File: modules/furlough/view/textlist.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    <th></th>
+  </tr>
+  <tr>
+    <td class="menu">ID</td>
+    <td class="menu">Text</td>
```
Comment (mladjo2505) [inline modules/furlough/view/textlist.php]: `tekst`

File: modules/workrelation/model/workrelation.class.php [hunk 6]
```diff
@@ -2,8 +2,46 @@
...
+require_once($_SETUP['HOME_DIR'] . "/code/lib/validation/validation.class.php");
+
 class work_relation {
 
+  public static function validate_furloughs($furloughs) {
+    $errors = array();
+
+    foreach ($furloughs as $furlough) {
+      $id = $furlough->FurloughID;
+      $error_prefix = "Permisjons og permitterings ". $id .": ";
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Should be `Permisjon og permittering`

File: modules/workrelation/model/workrelation.class.php [hunk 7]
```diff
@@ -2,8 +2,46 @@
...
+      if(self::empty_date($furlough->FurloughStop) || !self::valid_date($furlough->FurloughStop)) {
+        $errors[] = $error_prefix . "Slutt dato kan ikke v&aelig;re blank og m&aring; v&aelig;re en gyldig dato.";
+      }
+
+      if(!self::empty_date($furlough->FurloughStop) && $furlough->FurloughStop < $furlough->FurloughStart) {
+        $errors[] = $error_prefix . "Slutt dato kan ikke v&aelig;re f&aring;r start dato.";
+      }
+
+      if(self::empty_percent($furlough->FurloughPercent) || !self::valid_percent($furlough->FurloughPercent)) {
+        $errors[] = $error_prefix . "Prosent kan ikke v&aelig;re tom og m&aring; v&aelig;re st&oslash;rre enn 0 og 100 p&aring; det meste.";
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Should be `Prosent kan ikke v&aelig;re tom og m&aring; v&aelig;re mellom 0 og 100.`

File: modules/altinnsalary/model/altinnreport.class.php [hunk 8]
```diff
@@ -501,7 +524,85 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
...
+        ' p&aring; arbeidsforhold('.$furlough->WorkRelationID.')'.
+        ' permisjon('.$furlough->FurloughID.')', 'date');
+      // Error is: Furlough: Start date is not valid on self::fullNameForErrorMessage($employee) on work relation($furlough->WorkRelationID ) furlough( $furlough->FurloughID )
+      self::checkIfValid($furlough->FurloughStart,
+        'Permisjon: Startdato(' . $furlough->FurloughStart . ') er ikke gyldig for ' .
+        self::fullNameForErrorMessage($employee) .
+        ' p&aring; arbeidsforhold('.$furlough->WorkRelationID.')'.
+        ' permisjon('.$furlough->FurloughID.')', 'date');
+      $permisjon['startdato'] = $furlough->FurloughStart;
+
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php:549]: Check if the date is valid only if not empty. Same for end date.

**Comment by mladjo2505** [comment]
- [x] Add altinnsalary.altinn1list page to the maintenance page as a link.

File: modules/altinnsalary/model/altinnreport.class.php [hunk 9]
```diff
@@ -501,7 +524,85 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
...
       $arbeidsforhold['sisteDatoForStillingsprosentendring'] = strftime('%F', strtotime($last_change_of_work_percentage));
+
+      # Check if ther is any furlough and stuff
+    }
+
+    // select all furlough active in the report period
+    $query_furlough = "SELECT * FROM workrelationfurlough WHERE WorkRelationID = " . $work_relation->WorkRelationID . " AND (('" . $this->period . "-01' BETWEEN FurloughStart AND FurloughStop) OR ((YEAR(FurloughStart) = YEAR('" . $this->period . "-01') AND MONTH(FurloughStart) = MONTH('" . $this->period . "-01')) OR (YEAR(FurloughStop) = YEAR('" . $this->period . "-01') AND MONTH(FurloughStop) = MONTH('" . $this->period . "-01'))))";
+    $result_furlough = $_lib['db']->db_query($query_furlough);
+    while($furlough = $_lib['db']->db_fetch_object($result_furlough)) {
+      var_dump($furlough);
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: remove

File: modules/altinnsalary/model/altinnreport.class.php [hunk 10]
```diff
@@ -501,7 +524,85 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
...
+      // Error is: Furlough: Start date is not valid on self::fullNameForErrorMessage($employee) on work relation($furlough->WorkRelationID ) furlough( $furlough->FurloughID )
+      self::checkIfValid($furlough->FurloughStart,
+        'Permisjon: Startdato(' . $furlough->FurloughStart . ') er ikke gyldig for ' .
+        self::fullNameForErrorMessage($employee) .
+        ' p&aring; arbeidsforhold('.$furlough->WorkRelationID.')'.
+        ' permisjon('.$furlough->FurloughID.')', 'date');
+      $permisjon['startdato'] = $furlough->FurloughStart;
+
+      // Error is: Furlough: End date is missing on self::fullNameForErrorMessage($employee) on work relation($furlough->WorkRelationID ) furlough( $furlough->FurloughID )
+      self::checkIfEmpty($furlough->FurloughStop,
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: can be empty, but if not empty, must be valid

File: db/changes/137_create_table_work_relation_furlogh.sql [hunk 11]
```diff
@@ -0,0 +1,17 @@
+CREATE TABLE IF NOT EXISTS `workrelationfurlough` (
```
Comment (mladjo2505) [inline db/changes/137_create_table_work_relation_furlogh.sql:1]: - [x] Rename columns without Furlough prefix.

File: modules/altinnsalary/model/altinnreport.class.php [hunk 12]
```diff
@@ -501,7 +524,88 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
         self::checkIfEmpty($last_change_of_work_percentage, 'L&oslash;nnslipp: Mangler stillingsprosentendret p&aring; L' . $salary->JournalID, 'date');
       }
       $arbeidsforhold['sisteDatoForStillingsprosentendring'] = strftime('%F', strtotime($last_change_of_work_percentage));
+
+      # Check if ther is any furlough and stuff
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: nice comment

File: modules/altinnsalary/model/altinnreport.class.php [hunk 13]
```diff
@@ -501,7 +524,88 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
...
+        self::checkIfValid($furlough->FurloughStop,
+          'Permisjon: Sluttdato(' . $furlough->FurloughStop . ') er ikke gyldig for ' .
+          self::fullNameForErrorMessage($employee) .
+          ' p&aring; arbeidsforhold('.$furlough->WorkRelationID.')'.
+          ' permisjon('.$furlough->FurloughID.')', 'date');
+        $permisjon['stuttdato'] = $furlough->FurloughStop;
+      }
+
+      // Error is: Furlough: End date can't be before start date on self::fullNameForErrorMessage($employee) on work relation($furlough->WorkRelationID ) furlough( $furlough->FurloughID )
+      if (!$furlough_start_empty && !$furlough_end_empty && ($furlough->FurloughStop < $furlough->FurloughStart)) {
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: Should it be `<=`? For example one day absence.

File: modules/altinnsalary/model/altinnreport.class.php [hunk 14]
```diff
@@ -501,7 +524,88 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
         self::checkIfEmpty($last_change_of_work_percentage, 'L&oslash;nnslipp: Mangler stillingsprosentendret p&aring; L' . $salary->JournalID, 'date');
       }
       $arbeidsforhold['sisteDatoForStillingsprosentendring'] = strftime('%F', strtotime($last_change_of_work_percentage));
+
+      # Check if ther is any furlough and stuff
+    }
+
+    // select all furlough active in the report period
+    $query_furlough = "SELECT * FROM workrelationfurlough WHERE WorkRelationID = " . $work_relation->WorkRelationID . " AND (('" . $this->period . "-01' BETWEEN FurloughStart AND FurloughStop) OR ((YEAR(FurloughStart) = YEAR('" . $this->period . "-01') AND MONTH(FurloughStart) = MONTH('" . $this->period . "-01')) OR (YEAR(FurloughStop) = YEAR('" . $this->period . "-01') AND MONTH(FurloughStop) = MONTH('" . $this->period . "-01'))))";
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: Will between part of the query work for this example:
Sending for 2016-09.
Furlough is from 2016-05-23 and no end date.

File: modules/workrelation/model/workrelation.class.php [hunk 15]
```diff
@@ -2,8 +2,46 @@
...
+      if(self::empty_date($furlough->FurloughStart) || !self::valid_date($furlough->FurloughStart)) {
+        $errors[] = $error_prefix . "Start dato kan ikke v&aelig;re blank og m&aring; v&aelig;re en gyldig dato.";
+      }
+
+      if(self::empty_date($furlough->FurloughStop) || !self::valid_date($furlough->FurloughStop)) {
+        $errors[] = $error_prefix . "Slutt dato kan ikke v&aelig;re blank og m&aring; v&aelig;re en gyldig dato.";
+      }
+
+      if(!self::empty_date($furlough->FurloughStop) && $furlough->FurloughStop < $furlough->FurloughStart) {
+        $errors[] = $error_prefix . "Slutt dato kan ikke v&aelig;re f&aring;r start dato.";
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Edn date can be blank, but if not blank it needs to be a valid date.

File: modules/accountplan/view/employee.php [hunk 16]
```diff
@@ -445,11 +445,114 @@
     <td><? print $_lib['form3']->text(array('table'=>$db_table2, 'field'=>'WorkMeasurement', 'value'=>$work_relation->WorkMeasurement, 'pk'=> $WorkRelationID, 'class'=>'lodoreqfelt')) ?></td>
     <td><? print $_lib['form3']->date(array('table'=>$db_table2, 'field'=>'SalaryDateChangedAt', 'form_name' => 'work_relations', 'value'=>$work_relation->SalaryDateChangedAt, 'pk'=> $WorkRelationID, 'class'=>'lodoreqfelt')) ?></td>
     <td>
+
```
Comment (mladjo2505) [inline modules/accountplan/view/employee.php]: remove

File: modules/accountplan/view/employee.php [hunk 17]
```diff
@@ -445,11 +445,114 @@
...
+  <?
+    $db_table3 = 'workrelationfurlough';
+    $query_work_relation_furloghs = "SELECT * FROM $db_table3 WHERE WorkRelationID = $work_relation->WorkRelationID";
+    $result_work_relation_furloughs = $_lib['db']->db_query($query_work_relation_furloghs);
+    if ($_lib['db']->db_numrows($result_work_relation_furloughs)) {
+  ?>
+
+    <tr>
+
```
Comment (mladjo2505) [inline modules/accountplan/view/employee.php]: remove empty line

File: modules/altinnsalary/model/altinnreport.class.php [hunk 18]
```diff
@@ -66,14 +69,34 @@ function addReplacementMessageID($message_id) {
...
+
+/* Helper function to check if the variable is valid
+ * calls a sub function for for date, percent, ...
+ */
+  function checkIfValid($field, $error_message, $type = '') {
+    if ($type == 'date') $is_valid = validation::date($field);
+    elseif ($type == 'percent') $is_valid = self::isValidPercent($field);
+    else {
+      $error_message = 'Unknown type ' . $type;
+      $is_valid = true;
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: false?

File: modules/altinnsalary/model/altinnreport.class.php [hunk 19]
```diff
@@ -501,7 +524,88 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
...
+      self::checkIfEmpty($furlough->FurloughDescription,
+        'Permisjon: Mangler beskrivelse for ' .
+        self::fullNameForErrorMessage($employee) .
+        ' p&aring; arbeidsforhold('.$furlough->WorkRelationID.')'.
+        ' permisjon('.$furlough->FurloughID.')');
+      $permisjon['beskrivelse'] = $furlough->FurloughDescription;
+
+      $arbeidsforhold[]['permisjon'] = $permisjon;
     }
+
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: remove

File: modules/workrelation/model/workrelation.class.php [hunk 20]
```diff
@@ -2,8 +2,46 @@
 /* Work relation class
  */
 
+// include validation class
+require_once($_SETUP['HOME_DIR'] . "/code/lib/validation/validation.class.php");
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Would be nice to have a method and not this ugly code.

File: modules/workrelation/model/workrelation.class.php [hunk 21]
```diff
@@ -2,8 +2,46 @@
...
 
+  public static function validate_furloughs($furloughs) {
+    $errors = array();
+
+    foreach ($furloughs as $furlough) {
+      $id = $furlough->FurloughID;
+      $error_prefix = "Permisjon og permittering ". $id .": ";
+
+      if(self::empty_field($furlough->FurloughText)) {
+        $errors[] = $error_prefix . "Tekst kan ikke v&aelig;re blank.";
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php:20]: Add english text as comment

File: modules/workrelation/model/workrelation.class.php [hunk 22]
```diff
@@ -2,8 +2,46 @@
...
+      }
+
+      if(self::empty_date($furlough->FurloughStart) || !self::valid_date($furlough->FurloughStart)) {
+        $errors[] = $error_prefix . "Start dato kan ikke v&aelig;re blank og m&aring; v&aelig;re en gyldig dato.";
+      }
+
+      if(self::empty_date($furlough->FurloughStop) || !self::valid_date($furlough->FurloughStop)) {
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: End date can be blank, but if not blank it needs to be a valid date.

File: modules/workrelation/model/workrelation.class.php [hunk 23]
```diff
@@ -2,8 +2,46 @@
...
+        $errors[] = $error_prefix . "Slutt dato kan ikke v&aelig;re blank og m&aring; v&aelig;re en gyldig dato.";
+      }
+
+      if(!self::empty_date($furlough->FurloughStop) && $furlough->FurloughStop < $furlough->FurloughStart) {
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Start also needs to be not empty for this check.

File: modules/workrelation/model/workrelation.class.php [hunk 24]
```diff
@@ -92,6 +130,18 @@ public static function empty_field($value) {
     return empty($value) || $value == "0";
   }
 
+  public static function empty_percent($value) {
+    return empty($value) || $value == 0;
```
Comment (mladjo2505) [inline modules/workrelation/model/workrelation.class.php]: Should not check if the value is 0.

File: modules/altinnsalary/model/altinnreport.class.php [hunk 25]
```diff
@@ -502,6 +525,84 @@ function generateArbeidsforholdArray($salary, $employee, $work_relation, $use_on
       }
       $arbeidsforhold['sisteDatoForStillingsprosentendring'] = strftime('%F', strtotime($last_change_of_work_percentage));
     }
+
+    // select all furlough active in the report period
+    $query_furlough = "SELECT * FROM workrelationfurlough WHERE WorkRelationID = " . $work_relation->WorkRelationID . " AND (('" . $this->period . "-01' BETWEEN Start AND Stop) OR ((YEAR(Start) = YEAR('" . $this->period . "-01') AND MONTH(Start) = MONTH('" . $this->period . "-01')) OR (YEAR(Stop) = YEAR('" . $this->period . "-01') AND MONTH(Stop) = MONTH('" . $this->period . "-01'))))";
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: Will between part of the query work for this example:
Sending for 2016-09.
Furlough is from 2016-05-23 and no end date.

**Comment by sls-lnluksa** [comment]
First round review: Nikola Lukovic (sls-lnluksa)
Final review with: @stoivo 

@stoivo Check if Furlough can overlap
Example: Having two Furlough with same date, or partly overlapping dates.

**Comment by stoivo** [comment]
Migration 137  might already be used on master, so change to next free one.

File: modules/altinnsalary/model/altinnreport.class.php [hunk 26]
```diff
@@ -66,14 +69,34 @@ function addReplacementMessageID($message_id) {
     $this->erstatterMeldingsId = $message_id;
   }
 
+/* Helper function to check if the variable is a valid date
+ */
+  function isValidPercent($field) {
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: rename to `isValidFurloughPercent`

File: modules/altinnsalary/model/altinnreport.class.php [hunk 27]
```diff
@@ -66,14 +69,34 @@ function addReplacementMessageID($message_id) {
...
+    return is_numeric($field) && $field > 0 && $field <= 100;
+  }
+
+/* Helper function to check if the variable is valid
+ * calls a sub function for for date, percent, ...
+ */
+  function checkIfValid($field, $error_message, $type = '') {
+    if ($type == 'date') $is_valid = validation::date($field);
+    elseif ($type == 'percent') $is_valid = self::isValidPercent($field);
```
Comment (mladjo2505) [inline modules/altinnsalary/model/altinnreport.class.php]: rename to `isValidFurloughPercent` and type to `'furlough_percent'`

**Pull request merged**